Add boringcrypto image

dimitarvdimitrov opened the pull request:
This adds a boringcrypto image 

* builds and pushes amd64/arm64 boringcrypto image upon `make publish-images`
	* images are pushed to `grafana/rollout-operator-boringcrypto` 
	* also build the binary in an alpine container as opposed to a Debian. This means that the binary is linked against musl instead of glibc and running alpine with glibc may come with problems
* runs unit and integration tests with boringcrypto

dimitarvdimitrov commented:
The CI is failing the same bug that @andyasp encountered here https://github.com/grafana/rollout-operator/pull/70#issuecomment-1634544728

aknuds1 commented on Makefile:
```diff
@@ -13,22 +13,38 @@ GO_FILES := $(shell find . $(DONT_FIND) -o -type f -name '*.go' -print)
+
+.PHONY: publish-images
+publish-images: clean
+	docker buildx build --push --platform linux/amd64 --build-arg revision=$(GIT_REVISION) --build-arg BUILDTARGET=rollout-operator -t $(IMAGE_PREFIX)/rollout-operator:$(IMAGE_TAG) .
```
Should you no longer push linux/arm64?

aknuds1 commented on Makefile:
```diff
@@ -13,22 +13,38 @@ GO_FILES := $(shell find . $(DONT_FIND) -o -type f -name '*.go' -print)
-.PHONY: publish-image
-publish-image: clean
-	docker buildx build --push --platform linux/amd64,linux/arm64 --build-arg revision=$(GIT_REVISION) -t $(IMAGE_PREFIX)/rollout-operator:$(IMAGE_TAG) .
+.PHONY: build-image
```
Isn't the -boringcrypto suffix missing?
```suggestion
.PHONY: build-image-boringcrypto
```

dimitarvdimitrov replied:
I removed the phony altogether. I don't think we'll have a file called `build-image-boringcrypto` anyways. And IIRC we don't need .phony unless it can collide with a file name and a target

aknuds1 reviewed (commented):
See questions.

aknuds1 commented on Makefile:
```diff
@@ -13,22 +13,38 @@ GO_FILES := $(shell find . $(DONT_FIND) -o -type f -name '*.go' -print)
+.PHONY: publish-images
+publish-images: clean
+	docker buildx build --push --platform linux/amd64 --build-arg revision=$(GIT_REVISION) --build-arg BUILDTARGET=rollout-operator -t $(IMAGE_PREFIX)/rollout-operator:$(IMAGE_TAG) .
+	docker buildx build --push --platform linux/amd64,linux/arm64 --build-arg revision=$(GIT_REVISION) --build-arg BUILDTARGET=rollout-operator-boringcrypto -t $(IMAGE_PREFIX)/rollout-operator-boringcrypto:$(IMAGE_TAG) .
```
Based on the PR description, are you supposed to push _only_ AMD64 for boringcrypto?

dimitarvdimitrov replied:
yes, but I had mistakenly taken arm64 out of the wrong build command. Should be fixed now

aknuds1 commented on Makefile:
```diff
@@ -13,22 +13,37 @@ GO_FILES := $(shell find . $(DONT_FIND) -o -type f -name '*.go' -print)
 integration: integration/mock-service/.uptodate
 	go test -v -tags requires_docker -count 1 -timeout 1h ./integration/...
 
+.PHONY: integration
```
```suggestion
.PHONY: integration-boringcrypto
```

aknuds1 commented on Makefile:
```diff
@@ -13,22 +13,37 @@ GO_FILES := $(shell find . $(DONT_FIND) -o -type f -name '*.go' -print)
 
+.PHONY: integration
+integration-boringcrypto: integration/mock-service/.uptodate
+	go test -v -tags requires_docker -count 1 -timeout 1h ./integration/...
```
How do you enable boringcrypto for this test?

aknuds1 reviewed (commented):
Please see comments

aknuds1 commented on Makefile:
```diff
@@ -13,22 +13,37 @@ GO_FILES := $(shell find . $(DONT_FIND) -o -type f -name '*.go' -print)
+.PHONY: integration
+integration-boringcrypto: integration/mock-service/.uptodate
+	go test -v -tags requires_docker -count 1 -timeout 1h ./integration/...
+
```
Extraneous newline
```suggestion
```

dimitarvdimitrov commented:
thanks for the reviews @aknuds1, I think I addressed all of the issues. Sorry for the state of this PR; I should have reviewed it on my own a few times before opening.

aknuds1 commented on .github/workflows/ci.yaml:
```diff
@@ -35,6 +36,19 @@ jobs:
       - run: make build-image
       - run: make integration
 
+
```
[Nit] Extraneous newline
```suggestion
```

aknuds1 commented on .github/workflows/ci.yaml:
```diff
@@ -35,6 +36,19 @@ jobs:
+          check-latest: true
+      - run: make build-image-boringcrypto
+      - run: make integration
+
```
[Nit] Extraneous newline
```suggestion
```

aknuds1 commented on Makefile:
```diff
@@ -13,22 +13,36 @@ GO_FILES := $(shell find . $(DONT_FIND) -o -type f -name '*.go' -print)
 	GOOS=$(GOOS) GOARCH=$(GOARCH) CGO_ENABLED=0 go build -ldflags '-extldflags "-static"' ./cmd/rollout-operator
 
+rollout-operator-boringcrypto: $(GO_FILES)
+	GOEXPERIMENT=boringcrypto GOOS=$(GOOS) GOARCH=$(GOARCH) CGO_ENABLED=1 go build -ldflags '-extldflags "-static"' ./cmd/rollout-operator
```
Assuming that boringcrypto requires CGO.

dimitarvdimitrov replied:
yes, the linked openssl is in C and needs to called via CGO

dimitarvdimitrov replied:
there are more details here https://kupczynski.info/posts/fips-golang/ (is somewhat outdated because boringcrypto now supports arm64 as well as amd64, but the rest is still valid)

aknuds1 reviewed (approved):
LGTM, modulo a couple of newline nits.

dimitarvdimitrov commented:
I pushed a commit to pin the go version so we can see a passing CI. After that I'll undo the pin and merge on failing CI

dimitarvdimitrov commented:
I'm investigating the failing CI

dimitarvdimitrov commented:
I had to make some adjustment to the PR. As of now this is the amd4 binary built from this PR using the commit where CI tests ran - 7a0770d

```
$ file /tmp/rollout-operator
/tmp/rollout-operator: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, Go BuildID=GpWoX651AkwtQPsrgVNx/nuEX87cG0CpHr_FAQAlZ/WGnquJfCp-utqUX_5FNc/eePK3vlRL3Hn65mybSiG, with debug_info, not stripped
```

and includes boringcrypto & FIPS symbols as opposed to standard library symbols

```
$ go tool nm /tmp/rollout-operator | grep FIPS
  461b80 t BORINGSSL_FIPS_abort
  45da80 t FIPS_mode_set
  45da90 t FIPS_read_counter
  401490 T _cgo_d8b1bdd8e714_Cfunc__goboringcrypto_FIPS_mode
  45da70 t _goboringcrypto_FIPS_mode
  65bd40 T crypto/internal/boring._Cfunc__goboringcrypto_FIPS_mode.abi0
 2a89888 D crypto/internal/boring._cgo_d8b1bdd8e714_Cfunc__goboringcrypto_FIPS_mode
  5eadc0 T crypto/internal/boring/sig.FIPSOnly.abi0
 2a8ae30 D crypto/tls.defaultCipherSuitesFIPS
 2a8aeb0 D crypto/tls.defaultFIPSCurvePreferences
```

 The tests pass on https://github.com/grafana/rollout-operator/pull/71/commits/7a0770dea6f54c50328075874062f9ec3a3e26b2 with the go version pinned to 1.20.5. I reverted the pin in f4ef221a1041c8f073fb0c67bd9b368bacf23e99 that's why the branch is now failing CI. The failing CI is due to a bugfix in docker which hasn't reached us yet (see https://github.com/grafana/rollout-operator/pull/70#issuecomment-1634544728)

aknuds1 commented on Dockerfile:
```diff
@@ -1,14 +1,15 @@
-FROM --platform=$BUILDPLATFORM golang:1.20-bookworm AS build
```
What was the purpose of `--platform=$BUILDPLATFORM`? First time I see it. Was it unnecessary?

dimitarvdimitrov replied:
the `--platform` forced to use the host os/arch even for cross-compiling. I think I ran into problems with this when compiling into CGO because the toolchain for compiling cgo on amd64 wasn't available in the arm64 golang image. Removing the `--platform=$BUILDPLATFORM` is effectively using arm64 image to build the arm64 binary and amd64 image to build the amd64 binary.

aknuds1 replied:
Makes sense to me, thanks for the explanation.

aknuds1 commented on Dockerfile:
```diff
@@ -1,14 +1,15 @@
 
 FROM alpine:3.18
-RUN apk add --no-cache ca-certificates
+RUN apk add --no-cache ca-certificates libc6-compat gcompat
```
Are libc6-compat and gcompat necessary for boringcrypto linking to work?

Could you add a comment on why they're necessary?

aknuds1 commented on Makefile:
```diff
@@ -13,18 +13,28 @@ GO_FILES := $(shell find . $(DONT_FIND) -o -type f -name '*.go' -print)
+.PHONY: publish-images
+publish-images: clean
+	docker buildx build --push --platform linux/amd64,linux/arm64 --build-arg revision=$(GIT_REVISION) --build-arg BUILDTARGET=rollout-operator -t $(IMAGE_PREFIX)/rollout-operator:$(IMAGE_TAG) .
+	docker buildx build --push --platform linux/amd64,linux/arm64 --build-arg revision=$(GIT_REVISION) --build-arg BUILDTARGET=rollout-operator-boringcrypto -t $(IMAGE_PREFIX)/rollout-operator-boringcrypto:$(IMAGE_TAG) .
```
So boringcrypto works on ARM64 now?

dimitarvdimitrov replied:
boringcrypti has support for arm64, but I couldn't initially get the build going.

After making the binary dynamically linked and fixing the runtime dependencies arm64 works just as well

aknuds1 reviewed (commented):
Please see questions, otherwise good to go I think?

aknuds1 commented on Dockerfile:
```diff
@@ -1,14 +1,18 @@
 
 FROM alpine:3.18
-RUN apk add --no-cache ca-certificates
+# need gcompat for glibc compatibility of the boringcrypto image
+RUN apk add --no-cache ca-certificates gcompat
```
We don't need gcompat anymore :)
```suggestion
RUN apk add --no-cache ca-certificates
```

aknuds1 reviewed (approved):
LGTM, although you should remove the `gcompat` package in Alpine.

andyasp commented on Dockerfile:
```diff
@@ -1,11 +1,14 @@
-FROM --platform=$BUILDPLATFORM golang:1.20-bookworm AS build
+FROM golang:1.20-alpine AS build
```
```suggestion
FROM golang:1.20-alpine3.18 AS build
```
If you are linking libraries then specifying the minor alpine version matching the later stage may help prevent things from going wrong in the future if we temporarily drift from the latest version.

aknuds1 replied:
@andyasp good catch - I was just thinking about this fact in the meantime (my fault that we didn't specify Alpine version).

andyasp commented on CHANGELOG.md:
```diff
@@ -2,6 +2,7 @@
 
 ## main / unreleased
 
+* [FEATURE] Publish boringcrypto image for amd64. #71
```
```suggestion
* [FEATURE] Publish an additional boringcrypto image for linux/amd64,linux/arm64. #71
```
Adding the word "additional" to clarify that this doesn't affect the other image.

andyasp commented on CHANGELOG.md:
```diff
@@ -2,6 +2,7 @@
 ## main / unreleased
 
+* [FEATURE] Publish boringcrypto image for amd64. #71
 * [ENHANCEMENT] Update the intermediate build container for the Docker image to `golang:1.20-bookworm`. #66 #67
```
```suggestion
* [ENHANCEMENT] Update the intermediate build container for the Docker image to `golang:1.20-alpine3.18`. #66 #67 #71
```
You can tell by the 3 PRs that we've really made up our mind on this.

andyasp reviewed (approved):
Minor comments, nice job!

andyasp commented on Makefile:
```diff
@@ -13,18 +13,34 @@ GO_FILES := $(shell find . $(DONT_FIND) -o -type f -name '*.go' -print)
+.PHONY: publish-images
+publish-images: publish-standard-image publish-boringcrypto-image
+
+.PHONY: publish-standard image
```
```suggestion
.PHONY: publish-standard-image
```
This was my typo
